libnymea/coap/coappdu.cpp

```diff
@@ -504,70 +508,91 @@ QByteArray CoapPdu::pack() const
+                setPayload(QByteArray(payloadData, payloadLength));
+            }
+            return;
+        }
 
-    setToken(QByteArray((const char *)rawData + 4, tokenLength));
-    setStatusCode(static_cast<StatusCode>(rawData[1]));
-    setMessageId((qint16)data.mid(2,2).toHex().toUInt(0,16));
-
-    // parse options
-    int index = 4 + tokenLength;
-    quint8 optionByte = rawData[index];
-    quint16 delta = 0;
-    while (QByteArray::number(optionByte, 16) != "ff" && optionByte != 0) {
-        quint16 optionNumber = ((optionByte & 0xf0) >> 4);
-
-        // check option delta
-        if (optionNumber < 13) {
-            delta += optionNumber;
-        } else if (optionNumber == 13) {
-            // extended 8 bit option delta
-            delta += (quint8)(rawData[index + 1] + 13);
-            index += 1;
-        } else if (optionNumber == 14) {
-            // extended 16 bit option delta
-            delta += ((rawData[index + 1] << 8) | rawData[index + 2]) + 269;
-            index += 2;
-        } else if (optionNumber == 15) {
-            m_error = InvalidOptionDeltaError;
+        quint16 optionDelta;
+        optionDelta = (optionByte & 0xf0) >> 4;
+//        qCDebug(dcCoap()) << "Option delta:" << optionDelta;
+        quint16 optionLength = (optionByte & 0x0f);
+//        qCDebug(dcCoap()) << "Option length:" << optionLength;
+
+        if (optionDelta == 13) {
+            quint8 optionDeltaExtended;
+            stream >> optionDeltaExtended;
+            optionDelta = optionDeltaExtended + 13;
+//            qCDebug(dcCoap()).nospace() << "Extended option delta (8 bit): " << optionDelta << " (" << optionDeltaExtended << " + 13)";
+        } else if (optionDelta == 14) {
+            quint16 optionDeltaExtended;
+            stream >> optionDeltaExtended;
+            optionDelta = optionDeltaExtended + 269;
+//            qCDebug(dcCoap()).nospace() << "Extended option delta (16 bit): " << optionDelta << " (" << optionDeltaExtended << " + 269)";
         }
 
-        // check option length
-        quint16 optionLength = (optionByte & 0xf);
         if (optionLength == 13) {
-            // extended 8 bit option length
-            optionLength = (quint8)(rawData[index + 1] - 13);
-            index += 1;
+            quint8 optionLengthExtended;
+            stream >> optionLengthExtended;
+            optionLength = optionLengthExtended + 13;
+//            qCDebug(dcCoap()).nospace() << "Extended option length (8 bit): " << optionLength << " (" << optionLengthExtended << " + 13)";
         } else if (optionLength == 14) {
-            // extended 16 bit option delta
-            optionLength = ((rawData[index + 1] << 8) | rawData[index + 2]) - 269;
-            index += 2;
-        } else if (optionLength == 15) {
-            m_error = InvalidOptionLengthError;
+            quint16 optionLengthExtended;
+            stream >> optionLengthExtended;
+            optionLength = optionLengthExtended + 269;
```

Comment:
`-269` vs `+269` ?

Reply:
[According to the spec](https://en.wikipedia.org/wiki/Constrained_Application_Protocol#Option) yes. The value contained in the package, is the value - 269, thus, the 269 need to be added again here. Not sure if the old code did the correct thing with the `or` and `shift` oerators before subtracting, but in any case, wrong lengths/indexes in the old code have been the main reason for this change in the first place.